### PR TITLE
feat(acp): add session/list and session/fork support

### LINKED ISSUE #7978: feat(acp): add session/list, session/resume and session/fork support
Add support for ACP draft/unstable methods:

- `session/list` - List existing sessions with cursor-based pagination
- `session/fork` - Fork an existing session to create a new independent session

These are part of the ACP draft spec:
- https://agentclientprotocol.com/protocol/draft/schema#session-list
- https://agentclientprotocol.com/protocol/draft/schema#session-fork

Use cases:
- **session/list**: Clients can discover and load previous sessions
- **session/fork**: Create branches from existing conversations (e.g., for generating summaries without affecting original)

### PR DESCRIPTION
**Pull request by remorses**
Closes #7978

## Summary

- Add `session/list` (unstable) with cursor-based pagination
- Add `session/fork` (unstable) to fork existing sessions
- Update `@agentclientprotocol/sdk` to 0.12.0

Smoke test script: https://gist.github.com/remorses/ddec691b72cef0ae128dc440505596f6

### CODE REVIEW

File: packages/opencode/src/acp/agent.ts [hunk 1]
```diff
@@ -448,25 +457,76 @@ export namespace ACP {
...
+        cwd: session.directory,
+        title: session.title,
+        updatedAt: new Date(session.time.updated).toISOString(),
+      }))
+      const next = filtered.length > limit ? String(page[page.length - 1]?.time.updated ?? "") : undefined
+      const response: ListSessionsResponse = {
+        sessions: entries,
+      }
+      if (next) response.nextCursor = next
+      return response
```
[the remark this answers is not in the document]
Comment (remorses) [reply packages/opencode/src/acp/agent.ts]: Fixed: wrapped in try/catch with consistent error handling

File: packages/opencode/src/acp/agent.ts [hunk 2]
```diff
@@ -448,25 +457,76 @@ export namespace ACP {
...
+        const model = await defaultModel(this.config, directory)
+
+        const forked = await this.sdk.session
+          .fork(
             {
-              sessionID: sessionId,
+              sessionID: params.sessionId,
               directory,
             },
             { throwOnError: true },
```
[the remark this answers is not in the document]
Comment (remorses) [reply packages/opencode/src/acp/agent.ts:613]: Fixed: added replay of forked session history after fork

File: packages/opencode/src/acp/agent.ts [hunk 3]
```diff
@@ -448,25 +457,76 @@ export namespace ACP {
...
+        if (LoadAPIKeyError.isInstance(error)) {
+          throw RequestError.authRequired()
+        }
+        throw e
+      }
+    }
+
+    async unstable_listSessions(params: ListSessionsRequest): Promise<ListSessionsResponse> {
+      const input = params.cwd ? { directory: params.cwd } : undefined
+      const sessions = await this.sdk.session
```
[the remark this answers is not in the document]
Comment (remorses) [reply packages/opencode/src/acp/agent.ts]: Fixed: now checks if last exists before creating cursor string

File: packages/opencode/src/acp/agent.ts [hunk 4]
```diff
@@ -448,25 +457,76 @@ export namespace ACP {
...
+        title: session.title,
+        updatedAt: new Date(session.time.updated).toISOString(),
+      }))
+      const next = filtered.length > limit ? String(page[page.length - 1]?.time.updated ?? "") : undefined
+      const response: ListSessionsResponse = {
+        sessions: entries,
+      }
+      if (next) response.nextCursor = next
+      return response
+    }
```
[the remark this answers is not in the document]
Comment (remorses) [reply packages/opencode/src/acp/agent.ts:598]: Fixed: removed non-null assertion, now throws explicit error if data is missing

**Comment by nonsleepr** [comment]
Hey @remorses! 👋

I've been working on the same `session/list` feature and noticed your PR. Rather than duplicate efforts, I'd love to collaborate.

**What I have:**
- **113 tests** covering ACP functionality including `session/list` pagination edge cases
- Upgraded to **SDK 0.13.0** (vs 0.12.0 in this PR)

**One design difference to discuss:**

Your implementation loads all sessions then slices in memory:
```ts
const all = await Session.list({ cwd: ... })
return all.slice(start, start + limit)
```

My implementation uses backend pagination via the SDK's `start`/`limit` params:
```ts
const result = await Session.list({ cwd, start, limit })
```

The backend approach scales better for users with many sessions (doesn't load everything into memory).

**My branch:** https://github.com/nonsleepr/opencode/tree/feat/acp-session-list

**Happy to:**
1. Contribute my test suite to this PR
2. Help update to SDK 0.13.0
3. Discuss the pagination approach with you/maintainers

Let me know if you'd like to collaborate! If you're busy, I can also open a separate PR and let maintainers decide which approach to merge.

**Comment by remorses** [comment]
Hey @nonsleepr, thanks for the feedback! We updated the implementation to use cursor-based pagination following the ACP spec. Also added `session/resume` support.

Regarding your test suite - we took a look but decided not to include them since they rely heavily on mocks, which doesn't align with our testing approach.

We'll keep this PR moving forward. Feel free to review the latest changes!
